### DISCUSSION TITLE
Allow the usage of more modern C# APIs

### DISCUSSION THREAD

**Discussion by dotlogix** [godotengine/godot-proposals#5377] [Scripting]
So after the NET6 merge I asked my self why Godot does not provide any way to use the high performance APIs Net6 provides.

I am talking especially about the extensive usage of Godot.Array<T> instead of native T[] or Span<T>. The current approach means a lot of copying back and forth which slows down performance without any reason.
Would be nice if we as developers could optimize for performance in the hot paths and use the simpler APIs everywhere else.

As an example I could name the SurfaceTool or ComputeShaders.
These are used in more performance critical situations (I assume) so allowing more performant types being used would really help.

**Real world example: ComputeShaders**
I experimented a bit lately with compute shaders and it is a very inefficent API to use.

I want to use compute shaders to generate a lot of noise really, really fast.
I use storage buffers to pass data between the CPU and the GPU.

In the current implementation I would need to:
Create a float[] for some input values.
Create a byte[] and Buffer.BlockCopy the floats over to the byte[].
Then I can actually do sth with it and call StorageBufferCreate

Then execute the shader.

To get the results I then call BufferGetData which will allocate another byte[] internally and returns it.
Create a float[] for the output values.
Another Buffer.BlockCopy the bytes over to the output array

So we have a total of 4 allocations which means 4 times the memory consumption and putting more pressure to the GC.
You usually don't use ComputeShaders for very small and simple scenarios so these arrays can get very big.

With the usage of Span<T> we could Read and Write to the buffer using the original arrays and unsafe cast them to Span<byte> if even needed. We could even stackalloc them so no heap allocation at all.

**Comment by Calinou** [comment]
> I asked my self why Godot does not provide any way to use the high performance APIs Net6 provides.

The API needs to remain compatible with GDScript and GDExtension, which doesn't have Span types or anything like that.

**Reply by dotlogix** [reply]
Wouldn't it be possible to expose APIs from C++ using pointers?
You can also use Spans with pointers so the C# source generator could generate additional methods for convenience?

Unfortunately I don't really know how the generators work so not sure if that would be possible somehow?
